### DISCUSSION TITLE
Periodic Boundary Condition + Constant

### DISCUSSION THREAD

**Discussion by mtrautner** [jorgensd/dolfinx_mpc#76] [Ideas]
I'm looking to enforce a constraint like ` u[x,1] = u[x,0] + f0(t)` and `u[0,y] = u[1,y] + f1(t)` for some given functions f0 and f1 for all x and y on a 2D square. It seems like dolfinx_mpc is the closest to achieving such a constraint. Do you have an idea of which functions I might modify to make this happen? Thanks.

**Comment by jorgensd** [comment]
I did make a try at this a while back. It should work with very minor modifications, like:

https://github.com/jorgensd/dolfinx_mpc/blob/main/cpp/lifting.h#L260

Is there. It is only in lifting the constant becomes relevant, thus only minor modifications is needed. I'll see if I can find the mathematical derivation i did on this (late next week).

**Reply by mtrautner** [reply]
Okay thanks. If I have this right, the commented out lines 260~264 of what you linked subtract a constant times some matrix values from relevant entities of the linear system, so the idea would be to instead just add a constant in this spot?

**Reply by RFurze** [reply]
Were you able to get this modification working? I'm looking to implement precisely this kind of constraint in what sounds like a similar multiscale model.

**Comment by jorgensd** [comment]
I've been thinking more about this, and I am in the progress of designing a version that should support this.
To try to not modify the existing workflow too much, I'm suggesting the following design rule:
- A slave cannot be a DirichletBC, so all current constructors will still keep these inputs as they are.
  -  If one wants this behavior, one should create the constraint $u_i = g$ with MPC rather than DirichletBC.
- However, if $u_i = \alpha u_j + u_k, u_j = g$,$i\neq j$, $j\neq k$, $i \neq k$  this should be treated as $u_i \alpha  g + u_k$.

Mathematically, it breaks into the following problem:

$$
A x = b
$$

where $x$ is the unconstrained unknown, relating to the MPC as:

$$
K \bar x + \bar g
$$

Inserting this, we would get

$$
A K \bar x + A \bar g = b
$$

Symmetrizing and shuffling, we get:

$$
K^TAK \bar x = b - K^T A \bar g
$$

which is very similar to what we currently do in `dolfinx_mpc.apply_lifting` (there $\bar g$ is equal to the Dirichlet value when at a Dirichlet dof).
So I'm thinking the following API change:

`MultiPointConstraint` will take in a new, optional list of DirichletBCs.
What they will do effectively is to zero out the coefficients in front of the given degree of freedom
`MultiPointConstraint` will also take in a new array called `constants`, which would be equivalent to a `d_i` in the constraint `u_i = d_i`. 
In `apply_lifting`, the contributions from DirichletBC dofs and the extra constants would be accumulated.
